semantic-analysis/src/Analysis/Syntax.hs

```diff
@@ -104,15 +130,19 @@ letrec n m = do
 
 -- Parsing
 
-parseFile :: (Has (Throw String) sig m, MonadIO m) => FilePath -> m (File Term)
+parseFile :: (Has (Throw String) sig m, MonadIO m) => FilePath -> m (Source.Source, File Term)
 parseFile path = do
   contents <- liftIO (B.readFile path)
+  let sourcePath = replaceExtensions path "py"
+  -- FIXME: this is comprised of several terrible assumptions.
+  sourceContents <- Source.fromUTF8 . B.toStrict <$> liftIO (B.readFile sourcePath)
+  let span = decrSpan (Source.totalSpan sourceContents)
   case (A.eitherDecodeWith A.json' (A.iparse parseGraph) contents) of
     Left  (_, err)       -> throwError err
     Right (_, Nothing)   -> throwError "no root node found"
-    -- FIXME: this should get the path to the source file, not the path to the JSON.
-    -- FIXME: this should use the span of the source file, not an empty span.
-    Right (_, Just root) -> pure (File (Ref.fromPath (Path.absRel path)) root)
+    Right (_, Just root) -> pure (sourceContents, File (Reference (Path.absRel sourcePath) span) root)
```

Comment:
FWIW we’re planning to move away from typed paths (and to `OsPath` that comes with `filepath` 2.0), so if you have control over that `Reference` feel free to change it to just `FilePath`.